File: backend/channels.py
"""
ArenaPlay — PubNub Channel Name Helpers

All channel names for the ArenaPlay platform live here.
Changing a channel name in one place updates both backend and (via the JS copy) frontend.

Naming convention: dot-separated hierarchy → game.{gameId}.{purpose}
This enables future wildcard subscribe: game.abc123.*
"""


def questions(game_id: str) -> str:
    """Host publishes questions; all players subscribe. Persisted for catch-up."""
    return f"game.{game_id}.questions"


def answers_inbound(game_id: str) -> str:
    """
    All player answers land here. Functions onBefore validates and caps frequency.
    MVP: single shared inbound channel.
    Production upgrade path: game.{gameId}.answers.{playerId} per player with PAM regex.
    """
    return f"game.{game_id}.answers.inbound"


def leaderboard(game_id: str) -> str:
    """Backend publishes score updates after each round. Persisted for catch-up."""
    return f"game.{game_id}.leaderboard"


def reactions(game_id: str) -> str:
    """Players publish batched emoji bursts. Ephemeral — not persisted."""
    return f"game.{game_id}.reactions"


def chat(game_id: str) -> str:
    """Player chat. Functions onBefore moderates. Persisted for history."""
    return f"game.{game_id}.chat"


def crowd_energy(game_id: str) -> str:
    """Functions onAfter (reactions) publishes aggregated energy score. Ephemeral."""
    return f"game.{game_id}.crowd.energy"


def host_control(game_id: str) -> str:
    """Host publishes control events (start, reveal, end). Persisted for state sync."""
    return f"game.{game_id}.host.control"


def player_count(game_id: str) -> str:
    """Backend counter pattern: publishes {count: N} every 5s. Replaces native presence."""
    return f"game.{game_id}.player.count"


def player_notifications(player_id: str) -> str:
    """Backend publishes personalized alerts (rank change, bonus points). Persisted."""
    return f"player.{player_id}.notifications"


def illuminate_events(game_id: str) -> str:
    """Backend streams behavioral events to Illuminate for adaptive logic."""
    return f"illuminate.{game_id}.events"


# All channels a player should subscribe to for a given game
def player_subscribe_channels(game_id: str, player_id: str) -> list[str]:
    return [
        questions(game_id),
        leaderboard(game_id),
        reactions(game_id),
        chat(game_id),
        crowd_energy(game_id),
        host_control(game_id),
        player_count(game_id),
        player_notifications(player_id),
    ]


# All channels the host should subscribe to for a given game
def host_subscribe_channels(game_id: str) -> list[str]:
    return [
        answers_inbound(game_id),
        reactions(game_id),
        chat(game_id),
        crowd_energy(game_id),
        player_count(game_id),
    ]

File: frontend/src/channels.js
/**
 * ArenaPlay — PubNub Channel Name Helpers (JavaScript mirror of backend/channels.py)
 *
 * Single source of truth for channel names on the frontend.
 * Keep this in sync with backend/channels.py.
 */

export const ch = {
  questions:     (gameId) => `game.${gameId}.questions`,
  answersInbound:(gameId) => `game.${gameId}.answers.inbound`,
  leaderboard:   (gameId) => `game.${gameId}.leaderboard`,
  reactions:     (gameId) => `game.${gameId}.reactions`,
  chat:          (gameId) => `game.${gameId}.chat`,
  crowdEnergy:   (gameId) => `game.${gameId}.crowd.energy`,
  hostControl:   (gameId) => `game.${gameId}.host.control`,
  playerCount:   (gameId) => `game.${gameId}.player.count`,
  notifications: (playerId) => `player.${playerId}.notifications`,

  /** All channels a player subscribes to */
  playerChannels: (gameId, playerId) => [
    `game.${gameId}.questions`,
    `game.${gameId}.leaderboard`,
    `game.${gameId}.reactions`,
    `game.${gameId}.chat`,
    `game.${gameId}.crowd.energy`,
    `game.${gameId}.host.control`,
    `game.${gameId}.player.count`,
    `player.${playerId}.notifications`,
  ],

  /** All channels a host subscribes to */
  hostChannels: (gameId) => [
    `game.${gameId}.questions`,
    `game.${gameId}.answers.inbound`,
    `game.${gameId}.leaderboard`,
    `game.${gameId}.reactions`,
    `game.${gameId}.chat`,
    `game.${gameId}.crowd.energy`,
    `game.${gameId}.host.control`,
    `game.${gameId}.player.count`,
  ],
};

/** Event type constants — match backend models.py EventType */
export const EventType = {
  GAME_CREATED:       'GAME_CREATED',
  PLAYER_JOINED:      'PLAYER_JOINED',
  QUESTION_BROADCAST: 'QUESTION_BROADCAST',
  ANSWER_SUBMITTED:   'ANSWER_SUBMITTED',
  ANSWER_REVEAL:      'ANSWER_REVEAL',
  SCORES_UPDATED:     'SCORES_UPDATED',
  REACTION_BURST:     'REACTION_BURST',
  CHAT_MESSAGE:       'CHAT_MESSAGE',
  ROUND_COMPLETE:     'ROUND_COMPLETE',
  GAME_ENDED:         'GAME_ENDED',
  PLAYER_RECONNECTED: 'PLAYER_RECONNECTED',
  BONUS_TRIGGERED:    'BONUS_TRIGGERED',
  CROWD_ENERGY_UPDATE:  'CROWD_ENERGY_UPDATE',
  GAME_STARTED:         'GAME_STARTED',
  PLAYER_COUNT_UPDATE:  'PLAYER_COUNT_UPDATE',
};
